### ISSUE TITLE
`StopWhenUnneeded=true` ignored if the unit was spawned as dependency of a device unit

### ISSUE THREAD

**Issue by wucke13** [systemd/systemd#23410]
**systemd version the issue has been seen with**
 > systemd 249 (249.7)

**Used distribution**
 > NixOS 21.11

**Linux kernel version used** (`uname -a`)

 > `Linux ft-ssy-wznb 5.17.5 #1-NixOS SMP PREEMPT Wed Apr 27 12:41:17 UTC 2022 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > A unit with `StopWhenUnneeded=true` stops when `systemctl list-dependencies --reverse that-unit.service` yields only `that-unit.service`

**Unexpected behaviour you saw**
 > `that-unit.service` keeps running forever, even after the dependent which triggered `that-unit`s start becomes `inactive (dead)`

**Steps to reproduce the problem**

1. Add the following unit config and udev rule to the system
`that-unit.service`:
```ini
[Unit]
StopWhenUnneeded=true

[Service]
ExecStartPre=/bin/date
ExecStop=/bin/date
RemainAfterExit=true
Type=oneshot
```

`99-that-unit.rules`:
```udev
SUBSYSTEM=="usb", ACTION!="remove", ATTRS{idVendor}=="dead", ATTRS{idProduct}=="beef", TAG+="systemd", ENV{SYSTEMD_WANTS}+="that-unit.service"
```
2. Plug a device matching the udev rule in question. A device unit for it will become active, which triggers the start of`that-unit.service`. `systemctl list-dependencies --reverse that-unit.service` correctly lists the device unit(s) as dependents of `that-unit.service`.
3. Unplug the device. Its respective device units change from `active (plugged)` to `inactive (dead)`. `systemctl list-dependencies --reverse that-unit.service` now names no dependents of `that-unit.service`. Still, `systemctl status that-unit.service` lists `that-unit.service` as `active (exited)`. I would expect `that-unit.service` to be `inactive (dead)`, since `man systemd.unit` claims that `StopWhenUnneeded=true` `a unit will be automatically cleaned up if no other active unit requires it`. However, `that-unit.service` keeps `active (exited)`.

_Assumption 1_: The manpage speaks of `this unit will be stopped when it is no longer used`. I interpret this as when `systemctl list-dependencies --reverse` lists no other units, than the unit is no longer used
_Assumption 2_: The manpage speaks of `will be stopped`. I assume that this will be the equivalent in outcome to `systemctl stop`.

Possibly duplicate of https://github.com/systemd/systemd/issues/15676, however a new issue seems to be the right course of action here.

**Comment by arvidjaar** [comment]
Show full debug level log starting with inserting your USB device and until after you have unplugged it.

**Comment by wucke13** [comment]
[Here](https://github.com/systemd/systemd/files/8714020/log.txt) is a log. I `<REDACTED>` some personal information and changed the hostname, further more `that-unit.service` is called `udev-demo.service`. Sorry for the noise in the log, I already tried to end most services but still there is probably a lot of unrelated info in it

**Comment by arvidjaar** [comment]
I cannot reproduce it neither with v246 nor with v250. Your log shows multiple USB device that match your rule and try to trigger service. Did you check that all of them are dead from the systemd point of view?

**Comment by wucke13** [comment]
> I cannot reproduce it neither with v246 nor with v250.

That is quite unfortunate

> Your log shows multiple USB devices that match your rule and try to trigger service.

From where do you infer this? This is only one device, no hub or anything. Maybe it is just the raw USB device + the device through a driver?

> Did you check that all of them are dead from the systemd point of view?

Yes! See the following shell commands:

```console
$ systemctl status udev-demo.service; systemctl list-dependencies --reverse --all udev-demo.service
○ udev-demo.service
     Loaded: loaded (/etc/systemd/system/udev-demo.service; linked; vendor preset: enabled)
     Active: inactive (dead)
udev-demo.service
### echo Plugging in device
$ systemctl status udev-demo.service; systemctl list-dependencies --reverse --all udev-demo.service
● udev-demo.service
     Loaded: loaded (/etc/systemd/system/udev-demo.service; linked; vendor preset: enabled)
     Active: active (exited) since Thu 2022-05-19 08:55:43 CEST; 2s ago
    Process: 44223 ExecStartPre=/nix/store/ndkb1hkd009fch9dkzs31p9b7sa61s61-coreutils-9.0/bin/date (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
        CPU: 994us

May 19 08:55:43 ft-ssy-wznb systemd[1]: Starting udev-demo.service...
May 19 08:55:43 ft-ssy-wznb date[44223]: Thu May 19 08:55:43 AM CEST 2022
May 19 08:55:43 ft-ssy-wznb systemd[1]: Finished udev-demo.service.
udev-demo.service
● ├─sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1-1\x2d1:1.0.device
● └─sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
### Removing device
$ systemctl status udev-demo.service; systemctl list-dependencies --reverse --all udev-demo.service
● udev-demo.service
     Loaded: loaded (/etc/systemd/system/udev-demo.service; linked; vendor preset: enabled)
     Active: active (exited) since Thu 2022-05-19 08:55:43 CEST; 18s ago
    Process: 44223 ExecStartPre=/nix/store/ndkb1hkd009fch9dkzs31p9b7sa61s61-coreutils-9.0/bin/date (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
        CPU: 994us

May 19 08:55:43 ft-ssy-wznb systemd[1]: Starting udev-demo.service...
May 19 08:55:43 ft-ssy-wznb date[44223]: Thu May 19 08:55:43 AM CEST 2022
May 19 08:55:43 ft-ssy-wznb systemd[1]: Finished udev-demo.service.
udev-demo.service
```

From the second `list-dependencies` output I conclude that systemd does consider the device units as normal dependents, and lists them when they are there. From the third `list-dependencies` I conclude, that __no other dependents are active__ -- e.g. nobody needs the `udev-demo.service`.

**Comment by wucke13** [comment]
> I cannot reproduce it neither with v246 nor with v250.

That is quite unfortunate

> Your log shows multiple USB devices that match your rule and try to trigger service.

From where do you infer this? This is only one device, no hub or anything. Maybe it is just the raw USB device + the device through a driver?

> Did you check that all of them are dead from the systemd point of view?

Yes! See the following shell commands:

```console
$ systemctl status udev-demo.service; systemctl list-dependencies --reverse --all udev-demo.service
○ udev-demo.service
     Loaded: loaded (/etc/systemd/system/udev-demo.service; linked; vendor preset: enabled)
     Active: inactive (dead)
udev-demo.service
### echo Plugging in device
$ systemctl status udev-demo.service; systemctl list-dependencies --reverse --all udev-demo.service
● udev-demo.service
     Loaded: loaded (/etc/systemd/system/udev-demo.service; linked; vendor preset: enabled)
     Active: active (exited) since Thu 2022-05-19 08:55:43 CEST; 2s ago
    Process: 44223 ExecStartPre=/nix/store/ndkb1hkd009fch9dkzs31p9b7sa61s61-coreutils-9.0/bin/date (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
        CPU: 994us

May 19 08:55:43 my-hostname systemd[1]: Starting udev-demo.service...
May 19 08:55:43 my-hostname date[44223]: Thu May 19 08:55:43 AM CEST 2022
May 19 08:55:43 my-hostname systemd[1]: Finished udev-demo.service.
udev-demo.service
● ├─sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1-1\x2d1:1.0.device
● └─sys-devices-pci0000:00-0000:00:14.0-usb1-1\x2d1.device
### Removing device
$ systemctl status udev-demo.service; systemctl list-dependencies --reverse --all udev-demo.service
● udev-demo.service
     Loaded: loaded (/etc/systemd/system/udev-demo.service; linked; vendor preset: enabled)
     Active: active (exited) since Thu 2022-05-19 08:55:43 CEST; 18s ago
    Process: 44223 ExecStartPre=/nix/store/ndkb1hkd009fch9dkzs31p9b7sa61s61-coreutils-9.0/bin/date (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
        CPU: 994us

May 19 08:55:43 my-hostname systemd[1]: Starting udev-demo.service...
May 19 08:55:43 my-hostname date[44223]: Thu May 19 08:55:43 AM CEST 2022
May 19 08:55:43 my-hostname systemd[1]: Finished udev-demo.service.
udev-demo.service
```

From the second `list-dependencies` output I conclude that systemd does consider the device units as normal dependents, and lists them when they are there. From the third `list-dependencies` I conclude, that __no other dependents are active__ -- e.g. nobody needs the `udev-demo.service`.

**Comment by arvidjaar** [comment]
> From where do you infer this? 

Your own output shows two devices.

Try to make more precise rule to match just one device - is problem still reproducible?

**Comment by wucke13** [comment]
> Your own output shows two devices.

They are the same. My knowledge here is a bit shaky, but I think this is very typical for USB devices (I tried a different class of USB device and it exhibits the same behavior, two device units are allocated for it). I believe this is due to the fact that one USB device can register multiple interface descriptors at the same time. So the device unit with the shorter name for the USB device, and the longer name for the first (and only) (interface|Endpoint) descriptor in that device.

Please tell me, how I shall make my udev rule more specific, and what USB device you use that only creates a single device unit. I tried both a custom XHCI and a Storage device, and both allocate two device units.

**Comment by arvidjaar** [comment]
> They are the same.

They are different kernel device( path)s. Check udev properties; as far as I can tell, they have different `DEVTYPE`.

**Comment by wucke13** [comment]
> They are different kernel device( path)s. Check udev properties; as far as I can tell, they have different DEVTYPE.

Yes, one is a Device, the other an interface on that device.

```console
KERNEL[522.401864] add      /devices/pci0000:00/0000:00:14.0/usb2/2-3 (usb)
ACTION=add
DEVPATH=/devices/pci0000:00/0000:00:14.0/usb2/2-3
SUBSYSTEM=usb
DEVNAME=/dev/bus/usb/002/003
DEVTYPE=usb_device
PRODUCT=781/5588/100
TYPE=0/0/0
BUSNUM=002
DEVNUM=003
SEQNUM=4108
MAJOR=189
MINOR=130

KERNEL[522.402859] add      /devices/pci0000:00/0000:00:14.0/usb2/2-3/2-3:1.0 (usb)
ACTION=add
DEVPATH=/devices/pci0000:00/0000:00:14.0/usb2/2-3/2-3:1.0
SUBSYSTEM=usb
DEVTYPE=usb_interface
PRODUCT=781/5588/100
TYPE=0/0/0
INTERFACE=8/6/80
MODALIAS=usb:v0781p5588d0100dc00dsc00dp00ic08isc06ip50in00
SEQNUM=4109
```

Regardless of the terminology, I'm still wondering: why is it important whether this is a single or multiple kernel device( paths)s? As far as I understand, the condition for `StopWhenUnneeded` is `count(dependents) ==0`. So irrespective of whether 1, 2 or 100 device units `Want` the service with `StopWhenUneeded=true`, once all these device( unit)s are gone, the service should stop. I think I already highlighted that the service does not stop, even after `systemctl list-dependencies --reverse` hints that there are 0 dependents.

**Comment by janakj** [comment]
I can confirm what @wucke13 is seeing with systemd 247.3 on a Raspberry Pi (Debian 11 Bullseye). In my case, it appears to be happening only with systemd units that were activated as dependents of devices already present during boot.

Here is an example. I have a unit called joystick.service that looks as follows:
```
[Unit]
Description=Joystick driver
StopWhenUnneeded=true

[Service]
ExecStart=/joystick.sh
```
The unit is activated from the following udev rule:
```
SUBSYSTEMS=="usb", ACTION=="add", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c215", TAG+="systemd", ENV{SYSTEMD_WANTS}+="joystick.service"
```
I plug the joystick in while the Raspberry Pi is off, turn it on, and the unit is correctly activated after boot:
```
systemctl status joystick
● joystick.service - Joystick driver
     Loaded: loaded (/etc/systemd/system/joystick.service; static)
     Active: active (running) since Tue 2022-05-24 20:10:49 EDT; 4min 22s ago
   Main PID: 366 (joystick.sh)
      Tasks: 2 (limit: 1717)
        CPU: 2.267s
     CGroup: /system.slice/joystick.service
             ├─ 366 /bin/bash /joystick.sh
             └─1045 sleep 1

May 24 20:10:49 steve systemd[1]: Started Joystick driver.
```
The unit shows the following reverse dependencies:
```
systemctl list-dependencies --reverse joystick
joystick.service
● ├─sys-devices-platform-soc-3f980000.usb-usb1-1\x2d1-1\x2d1.3-1\x2d1.3:1.0-0003:046D:C215.0001-hidraw-hidraw0.device
● ├─sys-devices-platform-soc-3f980000.usb-usb1-1\x2d1-1\x2d1.3-1\x2d1.3:1.0-0003:046D:C215.0001-input-input1-event0.device
● ├─sys-devices-platform-soc-3f980000.usb-usb1-1\x2d1-1\x2d1.3-1\x2d1.3:1.0-0003:046D:C215.0001-input-input1-js0.device
● ├─sys-devices-platform-soc-3f980000.usb-usb1-1\x2d1-1\x2d1.3-1\x2d1.3:1.0-0003:046D:C215.0001-input-input1.device
● ├─sys-devices-platform-soc-3f980000.usb-usb1-1\x2d1-1\x2d1.3-1\x2d1.3:1.0.device
● └─sys-devices-platform-soc-3f980000.usb-usb1-1\x2d1-1\x2d1.3.device
```
So far so good.

Now I disconnect the USB device. The unit should be stopped since it has StopWhenUnneeded=true and its only dependants are the device units. But it is not the case, systemd left the unit running:
```
systemctl status joystick
● joystick.service - Joystick driver
     Loaded: loaded (/etc/systemd/system/joystick.service; static)
     Active: active (running) since Tue 2022-05-24 20:10:49 EDT; 7min ago
   Main PID: 366 (joystick.sh)
      Tasks: 2 (limit: 1717)
        CPU: 5.676s
     CGroup: /system.slice/joystick.service
             ├─ 366 /bin/bash /joystick.sh
             └─1472 sleep 1

May 24 20:10:49 steve systemd[1]: Started Shaky Joystick driver.
```
Even though the unit's list of dependants is empty now:
```
systemctl list-dependencies --reverse joystick
joystick.service
root@steve:/home/pi#
```

Furthermore, if I re-plug the USB device now and then unplug it again, systemd correctly stops the unit the second time. This leads me to believe that the problem depends on whether or not the device is plugged-in before boot.

**Comment by arvidjaar** [comment]
> why is it important whether this is a single or multiple kernel device( paths)s? 

This is just debugging aid. I cannot reproduce it and in my case there is single device that matches. If you cannot reproduce it with single device match, it may assist in finding offending code.

**Comment by wucke13** [comment]
@arvidjaar I changed the rule, adding `env{DEVTYPE}=="usb_device", `. Now only one device matches. The behavior remains the same:

```console
$ systemctl status that-unit.service; systemctl list-dependencies --reverse that-unit.service
○ that-unit.service
     Loaded: loaded (/etc/systemd/system/that-unit.service; linked; vendor preset: enabled)
     Active: inactive (dead)

May 28 21:37:12 zornpad date[87649]: Sat May 28 09:37:12 PM CEST 2022
May 28 21:37:12 zornpad systemd[1]: that-unit.service: Deactivated successfully.
May 28 21:37:12 zornpad systemd[1]: Stopped that-unit.service.
May 28 21:38:43 zornpad systemd[1]: Starting that-unit.service...
May 28 21:38:43 zornpad date[115540]: Sat May 28 09:38:43 PM CEST 2022
May 28 21:38:43 zornpad systemd[1]: Finished that-unit.service.
May 28 21:39:18 zornpad systemd[1]: Stopping that-unit.service...
May 28 21:39:18 zornpad date[115780]: Sat May 28 09:39:18 PM CEST 2022
May 28 21:39:18 zornpad systemd[1]: that-unit.service: Deactivated successfully.
May 28 21:39:18 zornpad systemd[1]: Stopped that-unit.service.
that-unit.service
$ systemctl status that-unit.service; systemctl list-dependencies --reverse that-unit.service
● that-unit.service
     Loaded: loaded (/etc/systemd/system/that-unit.service; linked; vendor preset: enabled)
     Active: active (exited) since Sat 2022-05-28 21:39:35 CEST; 100ms ago
    Process: 115802 ExecStartPre=/nix/store/47n5hzqpahs7yv84ia6cxp3jg9ca8r86-coreutils-9.0/bin/date (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
        CPU: 2ms

May 28 21:39:35 zornpad systemd[1]: Starting that-unit.service...
May 28 21:39:35 zornpad date[115802]: Sat May 28 09:39:35 PM CEST 2022
May 28 21:39:35 zornpad systemd[1]: Finished that-unit.service.
that-unit.service
● └─sys-devices-pci0000:00-0000:00:14.0-usb3-3\x2d1.device
$ systemctl status that-unit.service; systemctl list-dependencies --reverse that-unit.service
● that-unit.service
     Loaded: loaded (/etc/systemd/system/that-unit.service; linked; vendor preset: enabled)
     Active: active (exited) since Sat 2022-05-28 21:39:35 CEST; 6s ago
    Process: 115802 ExecStartPre=/nix/store/47n5hzqpahs7yv84ia6cxp3jg9ca8r86-coreutils-9.0/bin/date (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
        CPU: 2ms

May 28 21:39:35 zornpad systemd[1]: Starting that-unit.service...
May 28 21:39:35 zornpad date[115802]: Sat May 28 09:39:35 PM CEST 2022
May 28 21:39:35 zornpad systemd[1]: Finished that-unit.service.
that-unit.service
```

Also, this is on a newer systemd release and yet another computer:
```
systemd 250 (250.4)
+PAM +AUDIT -SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```
So, I'm really wondering why this is not reproducible at your side.

@janakj no matter how often I un-plug and re-plug the device, the service is started at the first time plugging in and remains from then on. So we have not the same behavior (but quite possibly both our observations are related to the same bug/problem).

**Comment by arvidjaar** [comment]
I tested with real USB device and in my case one of devices remains stuck. So while the end effect looks the same, it has clear cause (which itself is probably a different bug).

```
KERNEL[308.880810] unbind   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.0 (usb)
KERNEL[308.880987] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.0 (usb)
KERNEL[308.883170] unbind   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.1 (usb)
KERNEL[308.883213] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.1 (usb)
KERNEL[308.887007] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.2/net/enx0af69c4e9e7b/queues/rx-0 (queues)
KERNEL[308.887048] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.2/net/enx0af69c4e9e7b/queues/tx-0 (queues)
KERNEL[308.887081] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.2/net/enx0af69c4e9e7b (net)
KERNEL[308.896393] unbind   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.2 (usb)
KERNEL[308.896450] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:4.2 (usb)
KERNEL[308.896855] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/power_supply/apple_mfi_fastcharge/hwmon7 (hwmon)
KERNEL[308.897270] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/power_supply/apple_mfi_fastcharge/wakeup34 (wakeup)
KERNEL[308.897313] change   /devices/pci0000:00/0000:00:14.0/usb2/2-2/power_supply/apple_mfi_fastcharge (power_supply)
KERNEL[308.897338] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2/power_supply/apple_mfi_fastcharge (power_supply)
KERNEL[308.897372] unbind   /devices/pci0000:00/0000:00:14.0/usb2/2-2 (usb)
KERNEL[308.897402] remove   /devices/pci0000:00/0000:00:14.0/usb2/2-2 (usb)

bor@bor-Latitude-E5450:~$ systemctl -t device | grep usb
  sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.2.device                           loaded active plugged iPhone5/5C/5S/6                                                             
  sys-devices-pci0000:00-0000:00:1d.0-usb1-1\x2d1-1\x2d1.3-1\x2d1.3:1.0-bluetooth-hci0.device loaded active plugged /sys/devices/pci0000:00/0000:00:1d.0/usb1/1-1/1-1.3/1-1.3:1.0/bluetooth/hci0
bor@bor-Latitude-E5450:~$ 

июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.0.device: lost dependency Wants=that-unit.service
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: that-unit.service: lost dependency WantedBy=sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.0.device
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: that-unit.service: lost dependency ReferencedBy=sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.0.device
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.0.device: lost dependency References=that-unit.service
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.1.device: lost dependency Wants=that-unit.service
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: that-unit.service: lost dependency WantedBy=sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.1.device
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: that-unit.service: lost dependency ReferencedBy=sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.1.device
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.1.device: lost dependency References=that-unit.service
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.0.device: Changed plugged -> dead
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.0.device: Collecting.
июн 16 08:18:25 bor-Latitude-E5450 kernel: ipheth 2-2:4.2: Apple iPhone USB Ethernet now disconnected
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.1.device: Changed plugged -> dead
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.1.device: Collecting.
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.2-net-enx0af69c4e9e7b.device: Changed plugged -> dead
июн 16 08:18:25 bor-Latitude-E5450 systemd[1]: sys-devices-pci0000:00-0000:00:14.0-usb2-2\x2d2-2\x2d2:4.2-net-enx0af69c4e9e7b.device: Collecting.
```

So systemd seems to ignore the removal of USB interface for Ethernet. Device is not present in udev database anymoe.

**Comment by yuwata** [comment]
Thank you for the reports. Fix is waiting in #23848.
